libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2341,6 +2347,47 @@ bool TypeChecker::visit(IndexAccess const& _access)
 	return false;
 }
 
+bool TypeChecker::visit(IndexRangeAccess const& _access)
+{
+	_access.baseExpression().accept(*this);
+	TypePointer exprType = type(_access.baseExpression());
+	ArrayType const* arrayType = nullptr;
+
+	Expression const& start = _access.startExpression();
+	Expression const* end = _access.endExpression();
+
+	bool isLValue = false;
+	bool isPure = _access.baseExpression().annotation().isPure && start.annotation().isPure;
+
+	expectType(start, *TypeProvider::uint256());
+	if (end)
+	{
+		expectType(*end, *TypeProvider::uint256());
+		if (!end->annotation().isPure)
+			isPure = false;
+	}
+
+	if (exprType->category() == Type::Category::TypeType)
+	{
+		m_errorReporter.typeError(_access.location(), "Types cannot be sliced.");
+		_access.annotation().type = exprType;
+		return false;
+	}
+	else if (auto const* arraySlice = dynamic_cast<ArraySliceType const*>(exprType))
+		arrayType = &arraySlice->arrayType();
+	else if (!(arrayType = dynamic_cast<ArrayType const*>(exprType)))
+		m_errorReporter.fatalTypeError(_access.location(), "Index range access is only possible for arrays and array slices.");
```

Comment:
Should we only use one of the two expressions slice and range? Or are these different concepts for you?

Reply:
``IndexSliceAccess`` or ``SliceAccess`` both sound weird at least... and I'm not sure about ``ArrayRange`` instead of ``ArraySlice`` - to me an ``ArrayRange`` would rather just be a tuple of numbers, whereas the ``ArraySlice`` is the actual subarray given by a range... but the distinction is not too important to me - I'd be fine with changing it to one or the other, if you prefer.

Reply:
I'm only talking about the error messages for the users